### PR TITLE
[Feature] Implement Qwen3-ASR and wire it up as a seed-tts WER backend (#65)

### PR DESCRIPTION
**Pull request by 0xjeffro**
## Motivation

TTS benchmarks (and the TTS CI stages) score WER by transcribing the generated audio back to text with Whisper. This ASR pass is a major cost: Whisper's encoder effectively runs at batch size 1, so transcribing a full SeedTTS set is slow and is a meaningful chunk of TTS CI wall-clock time.

This PR (issue #65) integrates Qwen3-ASR into sglang-omni and wires it up as a WER backend for the seed-tts benchmark keeping accuracy on par with Whisper while cutting transcription time substantially, as the groundwork for switching the WER scorer (and speeding up the WER stages in CI).

## Modifications

**Qwen3-ASR model (served on sglang-omni)**
  - New `sglang_omni/models/qwen3_asr/`, ported from upstream sglang sgl-project/sglang#22073. Reuses sglang's existing `Qwen3ForCausalLM` decoder and `Qwen3OmniMoeAudioEncoder`, so it runs on our pinned sglang 0.5.8 without forking.
  - Registered in the omni model runner; serves the OpenAI-compatible `/v1/audio/transcriptions` endpoint. Verified on both 0.6B and 1.7B, EN + ZH.
  
**Benchmark integration**
  - `benchmark_tts_seedtts` gains `--qwen3-asr-port` to score WER via a running Qwen3-ASR server, alongside the existing local-Whisper / omni-Whisper-router backends. EN only for now.

## Related Issues

#65 
References upstream sgl-project/sglang#22073

## Accuracy Test

WER parity with the Whisper baseline (local, H200, EN, transcribe-only):
  - SeedTTS EN set, Higgs TTS audio (1088 samples): Qwen3-ASR-1.7B **~1.5%** vs
    Whisper **1.46%**.
  - The 20 SeedTTS reference clips used by CI stage-0: corpus WER **0.692%** (current Whisper stage-0 threshold 0.70%), passes stage-0's correctness bar locally.

## Benchmark & Profiling

Transcribe time on the SeedTTS EN set (1088 samples, local):
  - Whisper: 409s
  - Qwen3-ASR-1.7B (sequential) : 105s (~4x faster)
  - Qwen3-ASR-1.7B (client concurrency=4) : <60s 

## Checklist

- [x] Format your code according with pre-commit.
- [ ] Add unit tests.
- [ ] Update documentation / docstrings / example tutorials as needed.
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

## CI

CI is **not** switched to Qwen3-ASR in this PR. Moving stage-0 (and the TTS WER stages) off Whisper would require recalibrating their WER **and speed** thresholds (currently Whisper worst-of-5 baselines), so that's left as a follow-up for maintainers to decide.

### CODE REVIEW

File: benchmarks/eval/benchmark_tts_seedtts.py [hunk 1]
```diff
@@ -14,36 +14,22 @@
...
-    python -m sglang_omni.cli serve \
-        --model-path mistralai/Voxtral-4B-TTS-2603 \
-        --port 8000
-
-    3. For Higgs TTS:
-    python -m sglang_omni.cli serve \
-        --model-path boson-sglang/higgs-audio-v3-tts-4b-base \
-        --port 8000
-
-    # Full pipeline (generate + transcribe) — voice cloning
```
Comment (zhaochenyang20) [inline benchmarks/eval/benchmark_tts_seedtts.py:33]: why do you delete these?

File: benchmarks/eval/benchmark_omni_seedtts.py [hunk 2]
```diff
@@ -523,6 +523,13 @@ def _build_arg_parser() -> argparse.ArgumentParser:
         default="cuda:0",
         help="Legacy alias for --device (ASR transcription device).",
     )
+    parser.add_argument(
+        "--asr-router-port",
+        type=int,
+        default=None,
+        help="Qwen3-ASR sglang-omni router port for WER transcription "
+        "(/v1/audio/transcriptions). Required for EN; recommended for ZH.",
+    )
```
Comment (zhaochenyang20) [inline benchmarks/eval/benchmark_omni_seedtts.py:542]: As I said, no need to add a separate parameter of the `--asr-router-port`. Reuse Qwen3 Omni's server port.

File: benchmarks/eval/benchmark_tts_seedtts.py [hunk 3]
```diff
@@ -209,6 +209,122 @@ class TtsSeedttsBenchmarkConfig:
...
+        base_url=base_url,
+        log_path=log_path,
+        server_timeout=server_timeout,
+    )
+    try:
+        yield
+    finally:
+        logger.info(f"Stopping server ({model_path})")
+        _stop_server(proc)
+        _wait_for_gpu_idle()
```
Comment (zhaochenyang20) [inline benchmarks/eval/benchmark_tts_seedtts.py]: god damn fuck you do not understand how to reuse the exisiting functions in our repo? All these functions has already be implemented.

File: benchmarks/tasks/tts.py [hunk 4]
```diff
@@ -223,41 +223,65 @@ def load_omni_whisper_asr(
...
-        )
-        t0 = time.perf_counter()
-        pipe = pipeline(
-            "automatic-speech-recognition",
-            model="openai/whisper-large-v3",
-            chunk_length_s=30,
-            device=device,
+        raise ValueError(
+            "English WER transcription requires a Qwen3-ASR sglang-omni router "
+            "(DP=2). Start the router separately and pass asr_router_port."
```
Comment (zhaochenyang20) [inline benchmarks/tasks/tts.py]: DP is what CI is running on. Users may use other DP num.

Comment (zhaochenyang20) [reply benchmarks/tasks/tts.py]: What I said is, delete all the contents of `load_asr_model` with `transformers`. Purly use SGLang Omni Router to Launch ASR models. Default use Qwen3 ASR 1.6B, openai Whisper also can be used.

File: tests/test_model/omni_whisper_wer_utils.py [hunk 5]
```diff
@@ -1,35 +1,42 @@
...
     ManagedRouterHandle,
     launch_managed_router,
 )
 from tests.utils import wait_for_gpu_memory_release
 
-WHISPER_MODEL_PATH = "openai/whisper-large-v3"
-WHISPER_ASR_WORKER_ARGS = "--stages.0.factory-args.max-running-requests 1"
-WHISPER_ROUTER_STARTUP_TIMEOUT = 600
+ASR_MODEL_PATH = QWEN3_ASR_MODEL_PATH
+ASR_WORKER_ARGS = "--stages.0.factory-args.max-running-requests 1"
```
Comment (zhaochenyang20) [inline tests/test_model/omni_whisper_wer_utils.py]: Qwen3 ASR do not have this ` "--stages.0.factory-args.max-running-requests 1"` requirment. Delete it.

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

File: benchmarks/tasks/tts.py [hunk 6]
```diff
@@ -893,7 +918,8 @@ def run_seedtts_transcribe(
     asr = _resolve_asr_backend(
         config.lang,
         config.device,
-        whisper_router_port=whisper_router_port,
+        asr_router_port=asr_router_port,
+        asr_model_path=getattr(config, "asr_model_path", QWEN3_ASR_MODEL_PATH),
```
Comment (zhaochenyang20) [inline benchmarks/tasks/tts.py]: Never use a dynamic attribute.

**Comment by Hayden727** [comment]
### Qwen3-ASR speed check against the previous Whisper gate

Source: successful CI run on commit `8782d64`, before the rename-only commit `802cac2`: [TTS CI / stage 1 - Qwen3-ASR correctness](https://github.com/sgl-project/sglang-omni/actions/runs/26756279299/job/78873806608), [TTS CI / stage 2 - non-streaming](https://github.com/sgl-project/sglang-omni/actions/runs/26756279299/job/78873806600), and [TTS CI / stage 3 - streaming](https://github.com/sgl-project/sglang-omni/actions/runs/26756279299/job/78873837083). The latest commit `802cac2` only renames CI helpers/log labels and does not change the ASR request path.

The Qwen3-ASR CI keeps the previous Whisper-calibrated speed gate:

| Metric | Previous Whisper reference | CI threshold | Qwen3-ASR actual | Result |
| --- | ---: | ---: | ---: | --- |
| Throughput | 10.295 samples/s | >= 9.265 samples/s | 16.073 samples/s | faster, pass |
| Mean latency | 0.193 s | <= 0.213 s | 0.123 s | lower, pass |
| P95 latency | 0.555 s | <= 0.611 s | 0.164 s | lower, pass |
| Mean RTF | 0.0409 | <= 0.0450 | 0.0244 | lower, pass |
| P95 RTF | 0.1421 | <= 0.1563 | 0.0305 | lower, pass |

Full SeedTTS WER transcription stages also show the ASR request path is quick at `asr_concurrency=2`:

| Stage | Evaluated | Mean ASR latency | P95 ASR latency | ASR wall time | ASR throughput |
| --- | ---: | ---: | ---: | ---: | ---: |
| TTS non-streaming WER | 1088/1088 | 0.1037 s | 0.1225 s | 56.45 s | 19.27 samples/s |
| TTS streaming WER | 1088/1088 | 0.1032 s | 0.1224 s | 56.22 s | 19.35 samples/s |

Conclusion: by the existing speed thresholds and the latest completed CI logs, Qwen3-ASR is faster than the previous Whisper gate on the per-sample ASR path. If the overall CI stage wall-clock does not improve proportionally, the remaining time is mostly outside per-sample ASR transcription: server startup/lifecycle, TTS generation, artifact handling, WER bookkeeping, and SIM or UTMOS checks.

File: sglang_omni/models/qwen3_asr/configuration_qwen3_asr.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            trust_remote_code=trust_remote_code,
+        )
+        return cls(feature_extractor=feature_extractor, tokenizer=tokenizer)
+
+    def _get_feat_extract_output_lengths(self, input_lengths):
+        if not isinstance(input_lengths, torch.Tensor):
+            input_lengths = torch.tensor(input_lengths)
+        input_lengths_leave = input_lengths % 100
+        feat_lengths = (input_lengths_leave - 1) // 2 + 1
+        return ((feat_lengths - 1) // 2 + 1 - 1) // 2 + 1 + (input_lengths // 100) * 13
```
Comment (yxs) [inline sglang_omni/models/qwen3_asr/configuration_qwen3_asr.py]: Keep the audio-token-length formula single-source.

File: sglang_omni/models/qwen3_asr/request_builders.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+
+def _num_audio_tokens(num_mel_frames: int) -> int:
+    """Audio tokens emitted by the Qwen3-ASR encoder for a mel of given length.
+
+    Mirrors ``_get_feat_extract_output_lengths`` in the upstream modeling code
+    (3 stride-2 conv downsamples + windowed packing of 100-frame windows).
+    """
+    leave = num_mel_frames % 100
+    feat = (leave - 1) // 2 + 1
+    return ((feat - 1) // 2 + 1 - 1) // 2 + 1 + (num_mel_frames // 100) * 13
```
Comment (yxs) [inline sglang_omni/models/qwen3_asr/request_builders.py]: Keep the audio-token-length formula single-source.

File: sglang_omni/models/qwen3_asr/request_builders.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+            prompt_token_ids=input_ids,
+            max_new_tokens=request_max_new_tokens,
+            temperature=temperature,
+            audio_duration_s=audio_duration_s,
+            language=str(params.get("language") or "en"),
+            engine_start_s=time.perf_counter(),
+            stage_payload=payload,
+        )
+
+    def result_adapter(data: Qwen3ASRRequestData) -> StagePayload:
```
Comment (yxs) [inline sglang_omni/models/qwen3_asr/request_builders.py:280]: Looks like fragile.
The encode→decode round-trip isn't identity under byte-level BPE, leading/trailing whitespace and some Unicode can be rewritten, which feeds straight into WER.

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

**Review by zhaochenyang20 — APPROVED** [review]
